File: IT/Programmazione funzionale/Esercizi/C++/esercizio25.cpp
/*
file esercizio25.cpp
compilare con: g++ -std=c++11 esercizio25.cpp
*/

#include <iostream>
#include <functional>

using std::cout;
using std::endl;
using std::bind;
using namespace std::placeholders;

int main() {
	auto f = [] (const int a, const int b) { return a + b; };

	cout << "3 + 3 = " << f(3, 3) << endl;

	auto g = bind(f, 0, _1);

	cout << "id(3) = " << g(3) << endl;

	auto h = [] (const int a, const int b) { return a - b; };

	cout << "3 - 2 = " << h(3, 2) << endl;

	// calcola l'elemento inverso

	auto i = bind(h, 0, _1);

	cout << "Inverso additivo di 3 = " << i(3) << endl;
}

File: IT/Programmazione funzionale/Esercizi/Java/Esercizio25.java
/* file Esercizio25.java */

import java.util.function.Function;

public class Esercizio25 {
	public static void main(String [] args) {
		Function<Integer, Function<Integer, Integer>> f = a -> b -> a + b;

		System.out.println("3 + 3 = " + f.apply(3).apply(3));

		Function<Integer, Integer> g = f.apply(0);

		System.out.println("id(3) = " + g.apply(3));

		Function<Integer, Function<Integer, Integer>> h = a -> b -> a - b;

		System.out.println("3 - 2 = " + h.apply(3).apply(2));

		// calcola l'elemento inverso

		Function<Integer, Integer> i = h.apply(0);

		System.out.println("Inverso additivo di 3 = " + i.apply(3));
	}
}

